src/if_statements.py

```diff
@@ -570,6 +603,216 @@ def try_make_if_condition(
+        if node in visited_nodes or node == end or node == default_node:
+            continue
+        visited_nodes.add(node)
+        block_info = get_block_info(node)
+
+        if node != start and block_info.statements_to_write():
+            # Unless the node is the start node, it cannot have any statements to write
+            pass
+
+        elif is_empty_goto(node, end):
+            # Empty returns/gotos are special cases: the compiler may have folded
+            # this node together with one outside the original interval. So, this node
+            # may fail the `start not in node.dominators` check below.
+            # Otherwise, treat these like empty BasicNodes.
+            nodes_to_mark_emitted.add(node)
+            if isinstance(node, BasicNode):
+                node_queue.append(node.successor)
+            continue
+
+        elif start not in node.dominators or end not in node.postdominators:
+            # The node must be within the [start, end] interval
+            pass
+
+        elif isinstance(node, BasicNode):
+            # This node has no statements, so it is just a goto
+            nodes_to_mark_emitted.add(node)
+            node_queue.append(node.successor)
+            continue
+
+        elif isinstance(node, ConditionalNode):
+            # If this is a "switch guard" `if` statement, continue iterating on both branches
+            control_expr = switch_guard_expr(node)
+            if (
+                control_expr is not None
+                and early_unwrap_ints(control_expr) == uw_var_expr
+            ):
+                node_queue.append(node.fallthrough_edge)
+                node_queue.append(node.conditional_edge)
+                nodes_to_mark_emitted.add(node)
+                continue
+
+            # Check if the branch_condition is a comparison between var_expr and a Literal
+            assert block_info.branch_condition is not None
+            cond = simplify_condition(block_info.branch_condition)
+            if (
+                isinstance(cond, BinaryOp)
+                and isinstance(cond.right, Literal)
+                and early_unwrap_ints(cond.left) == uw_var_expr
+            ):
+                # IDO typically uses `x == N` and `x < N` patterns in these if trees, but it
+                # will use `x != N` when it needs to jump backwards to an already-emitted block.
+                # GCC will freely use other types of comparisons
+                # Examples from PM: func_8026E558, pr_load_npc_extra_anims
+                if cond.op == "==":
+                    if cond.right.value in cases:
+                        return None
+                    cases[cond.right.value] = node.conditional_edge
+                    node_queue.append(node.fallthrough_edge)
+                elif cond.op == "!=" and (
+                    node.block.index > node.conditional_edge.block.index
+                    or context.options.compiler == context.options.CompilerEnum.GCC
```

Comment:
do we end up with false positives if we remove this compiler check?

Reply:
Yeah, an example is [this if/else chain from `z_vimode` in OOT](https://github.com/zeldaret/oot/blob/master/src/code/z_vimode.c#L81-L105) and [this if/else chain in `z_boss_ganon`](https://github.com/zeldaret/oot/blob/826e5b71d11a60ccfbbe1267fbb4c93f7f4b080a/src/overlays/actors/ovl_Boss_Ganon/z_boss_ganon.c#L4622-L4799).